wellcomeml/datasets/winer.py

```diff
@@ -0,0 +1,222 @@
+
+# def load_winer():
+import tarfile
+import os
+from tqdm import tqdm 
+import json
+import random
+
+def create_train_test(
+    NE_path, vocab_path, docs_path,
+    train_processed_path, test_processed_path,
+    n_sample, prop_train, rand_seed=42
+    ):
+    """
+    NE_path: File path to the named entities
+    vocab_path: File path to the vocab dictionary
+    docs_path: File path to the documents texts (words coded according to the vocab)
+    train_processed_path: Output file path for the training data
+    test_processed_path: Output file path for the testing data
+    n_sample: Number of files to get data from, each file contains about 1000 documents
+    prop_train: Proportion of the file sample that will be in the training data
+
+    Utilise the three datasources to get the entities from a sample
+    of documents and save to a training and testing dataset files
+    in the form:
+
+    ID 12345
+    John Smith PERSON
+    rejects O
+    German LOC
+    call O
+    
+    Next 0
+    sentence 0
+
+    ID 444
+    Next
+    document
+    """
+
+    random.seed(rand_seed)
+    n_train = round(prop_train*n_sample)
+
+    print("Creating entities dictionary from {}".format(NE_path))
+    entities = {}
+    with tarfile.open(NE_path, "r:bz2") as tar:
+        for member in tqdm(tar.getmembers()):
+            f = tar.extractfile(member)
+            if f:
+                content = f.read().decode('utf-8', errors='ignore')
+                articles = content.replace('\t', ' ').split('ID ')
+                if articles[0]=='':
+                    articles = articles[1:]
+                for article in articles:
+                    article_entities = article.split('\n')
+                    article_id = article_entities[0]    
+                    if article_entities[-1]=='':
+                        article_entities = article_entities[1:-2]
+                    else:
+                        article_entities = article_entities[1:]
+                    entity_info = [l.split(' ') for l in article_entities]
+                    sentence_entities = {}
+                    for sentence_id, begin, end, ent_type in entity_info:
+                        sentence_id, begin, end, ent_type = int(sentence_id), int(begin), int(end), int(ent_type)
+                        if sentence_entities.get(sentence_id):
+                            sentence_entities[sentence_id].append([begin, end, ent_type])
+                        else:
+                            sentence_entities[sentence_id] = [[begin, end, ent_type]]
+                    entities[article_id] = sentence_entities
+
+    # Create id2word from vocab file (line number is id)
+    print("Creating id2word dictionary from {}".format(vocab_path))
+    id2word = {}
+    with open(vocab_path, "r") as vocab:
+        f = vocab.read()
+        lines = f.split('\n')
+        for i, line in enumerate(lines):
+            id2word[i] = line.split(' ')[0]
```

Comment:
A minor remark: This dictionary could be simply a list, since you are just attributing numbers from `0` to `len(lines)`. Then the entries could be accessed the same way by `id2word[i]`

Reply:
good point!